File: Matrix01_BFS.java
/* Time Complexity : O(m*n)
 * Space Complexity : O(m*n)
 * Did this code successfully run on Leetcode : Yes
 * Any problem you faced while coding this : No
*/

import java.util.LinkedList;
import java.util.Queue;

class Solution {
    public int[][] updateMatrix(int[][] matrix) {
        //null
        if(matrix == null || matrix.length == 0) return matrix;
        Queue<int []> q = new LinkedList<>();
        int m = matrix.length;
        int n = matrix[0].length;
        int [][] dirs = new int [][] {{0,1}, {0,-1}, {-1,0}, {1,0}};
        for(int i = 0; i < m ; i++){
            for(int j = 0; j < n; j++){
                if(matrix[i][j] == 0){
                    q.add(new int[] {i, j});
                } else {
                   matrix[i][j] = -1; 
                }
            }
        }
        int dist = 1; // dist = 0
        while(!q.isEmpty()){
            int size = q.size();
            //processing level
            for(int i = 0; i < size; i++){
                int [] curr = q.poll();
                for(int [] dir: dirs){
                    int nr = dir[0] + curr[0];
                    int nc = dir[1] + curr[1];
                    //check bounds
                    if(nr >= 0 && nc >= 0 && nr < m && nc < n && matrix[nr][nc] == -1){
                        matrix[nr][nc] = dist; //dist + 1 - if dist = 0
                        q.add(new int [] {nr, nc});
                    }
                }
            }
            dist++;
        }
        return matrix;
    }
}

File: matrix01-BFS.py
"""BFS with distance array"""
#Accepted by leetcode
#Time complexity - O(n) as we are visting every element
#Space complexity - O(n) as we are using queue and distance array

class Solution(object):
    def updateMatrix(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[List[int]]
        """
        if matrix == None or len(matrix) == 0:
            return matrix
        dist = [[0 for x in range(len(matrix[0]))] for y in range(len(matrix))]
        
        def bfs(matrix,x,y):
            q = []
            q.append((x,y))
            dirs = [(0,1),(1,0),(0,-1),(-1,0)]
            dist = 0
            while len(q) > 0:
                size = len(q)
                for i in range(size):
                    curr = q.pop(0)
                    for d in dirs:
                        r = d[0] + curr[0]
                        c = d[1] + curr[1]
                        if 0 <= r < len(matrix) and 0 <= c < len(matrix[0]):
                            if matrix[r][c] == 0:
                                return dist + 1
                            else:
                                q.append((r,c))
                dist += 1              
            return -1
        
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                if matrix[i][j] != 0:
                    dist[i][j] = bfs(matrix,i,j)
        return dist